### DISCUSSION TITLE
Cannot connect to terraria server (tshock) with public IP

### DISCUSSION THREAD

**Discussion by fetanq** [Pryaxis/TShock#1931] [Q&A] [answered]
Hey guys,

I have a Raspberry Pi running Tshock. All ports are configured to 7777.
When connecting from my PC to the terraria server using the private IP address of the Pi, everything works.
However, connecting with my public IP won't work. I don't even get a 'IP is connecting...' message in the server console. A friend also tried with no success.

Port forwarding is enabled on my router and the port is open (verified with yougetsignal.com). When verifying if the port is open, an entry pops up in the TShock console (198.199.98.246:47893 is connecting...).
I've (for testing purposes) also tried setting my pi as a DMZ host in my router to see if there is something blocked by the router with no success.
The Pi has a static IP and no other processes are running on port 7777.
I also have a webserver running on the same pi (internal port 80, external port 8080) which works.

Does someone have any ideas on what I can do to troubleshoot further?

Port forwarding configuration:
Protocol: TCP-UDP
External Port Start: 7777
External Port End: 7777
Internal Port: 7777
Internal Host: Private IP of the Pi
Remote Host: Any

Raspberry Pi 4 - Raspbian Buster
Tshock version TShock_4.4.0_Pre8_Terraria1.4.0.4

**Comment by kubedzero** [comment, marked as answer]
You say "connecting with my public IP won't work. I don't even get a 'IP is connecting...' message in the server console" and then you say "When verifying if the port is open, an entry pops up in the TShock console (198.199.98.246:47893 is connecting...)"

This verifies that the port forwarding is set up correctly since the port testing website is able to ping and reach all the way to the Tshock executable running on your Pi.

My guess is that your ISP disallows hosting a web server and has some prevention mechanism to keep other people's public IPs from pinging you. The reason the port test website might have worked is that the IP might be whitelisted as part of AWS, Google Cloud, Azure, etc. In other words, it might be that your ISP blocks direct traffic from the IP of your cell phone's hotspot, or the public IP of your friend's computer. 

I've had this issue myself in the past, and had to find a different internet provider to work around this. Another option would be to use some sort of a proxy service, or host the server cheaply in the cloud. 

Can you share how you tried to test connecting to your public IP? Cell phone hotspot, public library, etc?

**Reply by fetanq** [reply]
I tried connecting to the public IP with 3 different methods:
- Using the public IP in Terraria on my PC connected to my local network
- Using the public IP in terraria on my laptop connected to my phone's 4G
- A friend using Terraria on his PC outside of my network

I don't think it's the ISP though, Since the webserver to the same Raspberry Pi is working. 
Also I have set up a Terraria server in the past on a Raspberry Pi 2 and that one was working (that one was a few years ago, and I'm atill with the same ISP.
In the management console of my ISP's account there is an option to select "low security", but that only opens ports 80, 22 and 443. All other ports should be open by default.

Could the Pi itself block the connection?
It is a fresh install of the latest Raspbian. The first (and only) thing I did after the install was to install Pihole and Tshock.
Or some other steps that I can do to verify what is blocking the connection.

**Reply by kubedzero** [reply]
> using the private IP address of the Pi, everything works

If you're able to access the Pi normally from inside your network, the Pi is almost definitely not the problem. TShock doesn't have any sort of IP region/country based blocking. This means it cannot differentiate between an IP on your home network and an IP from the greater internet. If you can access TShock using a computer on your LAN, I'm confident that the Pi/Raspbian and TShock are set up correctly.

That would only leave your router and your ISP as links in the chain between an external IP connecting and TShock. We can only really take you at your word that your router is configured correctly, and the same goes for the ISP setup.

Why not experiment with TShock being port forwarded on one of the low security ports that your ISP specifically calls out? You mention "I also have a webserver running on the same pi which works" so maybe you could try shutting down the webserver for now and then running TShock on the same port the webserver was previously serving traffic. If that does not work, then there must be something with the Pi/TShock. If it does work, then that means there's a router/ISP/port issue.

**Reply by fetanq** [reply]
The issue has been solved. Thank you!
My Pihole server was interfering with Tshock. I didn't expect this and do not know why.
For some reason, running Pihole and Tshock breaks Tshock. Disabling Pihole not work. It has to not even be installed.
I'll install Pihole on an old Raspberry Pi 2 I have laying around

A fresh install of the Pi with a Tshock install afterwards worked. My friends can connect and life is good.
